packages/cozy-client/src/helpers.js

```diff
@@ -75,3 +75,139 @@ export const generateWebLink = ({
+ *
+ * @param {object} url          Object of URL elements
+ * @param {string} url.protocol Protocol to use in the origin (e.g. http)
+ * @param {string} url.hostname Hostname to use in the origin (e.g. claude.mycozy.cloud)
+ * @param {string} url.port     Port to use in the origin (e.g. 8080)
+ *
+ * @returns {string} Generated Well Known password change URL string
+ */
+const wellKnownUrl = url => uri(url) + '/.well-known/change-password'
+
+/* isValidOrigin - Checks whether a given URL is a valid Cozy origin
+ *
+ * This method tries to fetch the Well Known change password page of the Cozy
+ * supposedly at the given origin. This allows us to determine whether the given
+ * origin is the root URL of a Cozy or not via the status of the response:
+ * - a 200 response status means there's an actual Well Known password change
+ *   page accessible from the given origin so we suppose it's a valid Cozy
+ *   origin (i.e. it could be another site altogether though)
+ * - a 401 response status means the pointed page requires authentication so the
+ *   origin is probably pointing to a Cozy app
+ * - another status means there aren't any Cozy behind to the given origin
+ *
+ * @param {object} url          Object of URL elements
+ * @param {string} url.protocol Protocol to use in the origin (e.g. http)
+ * @param {string} url.hostname Hostname to use in the origin (e.g. claude.mycozy.cloud)
+ * @param {string} url.port     Port to use in the origin (e.g. 8080)
+ *
+ * @returns {Promise<boolean>} True if we believe there's a Cozy behind the given origin
+ * @throws {InvalidCozyUrlError} Thrown when we know for sure there aren't any Cozy behind the given origin
+ */
+const isValidOrigin = async url => {
+  const { status } = await fetch(wellKnownUrl(url))
+
+  if (status === 404) {
+    throw new InvalidCozyUrlError(url)
+  }
+  return status === 200
+}
+
+/**
+ * rootCozyUrl - Get the root URL of a Cozy from more precise ones
+ *
+ * The goal is to allow users to use any URL copied from their browser as their
+ * Cozy URL rather than trying to explain to them what we expect (e.g. when
+ * requesting the Cozy URL to connect an app).
+ * If we can't get the root URL either because there's no Cozy or the domain
+ * does not exist or anything else, we'll throw an InvalidCozyUrlError.
+ * Also, since we communicate only via HTTP or HTTPS, we'll throw an
+ * InvalidProtocolError if any other protocol is used.
+ *
+ * @param {URL} url The URL from which we'll try to get the root Cozy URL
+ *
+ * @returns {Promise<URL>} The root Cozy URL
+ */
+export const rootCozyUrl = async url => {
+  if (!['http:', 'https:'].includes(url.protocol)) {
+    throw new InvalidProtocolError(url)
+  }
+
+  // If the entered URL is good, use it
+  if (await isValidOrigin(url)) {
+    return url
+  }
+
+  // If the entered URL's lowest sub-domain contains a dash, remove it and
+  // what follows and try the new resulting url.
+  if (/^[^.-][^.]+-[^.-]+\./.test(url.hostname)) {
```

Comment:
Why the regexp is not simpler? Like `/^[^.-]+-[^.-]/

Reply:
The first version looked like this but it would not match urls like `https://jean-michel-drive.mycozy.cloud` (i.e. it only allows one dash in the sub-domain).
There might be a simpler version though but as always with Regexp, it's easy to miss it. I'll take a second look and see if I can come up with a simpler one.

Reply:
I couldn't figure out a simpler version. I suppose we can always change it later.